### PR TITLE
vmm: memory_manager: fix the issue virtio-mem start_addr is not align…

### PR DESCRIPTION
**Pull request by uMartinXu**
…ed with block

We can't assume virtio-mem block size is always the power of 2.
And the bitwise operation fixed by this commit can't align the start_addr
with block size if VIRTIO_MEM_DEFAULT_BLOCK_SIZE is not the power of 2.

Signed-off-by: MartinXu <martin.xu@intel.com>

### CODE REVIEW

**Comment by uMartinXu** [comment]
block size can't be changed by user on the runtime. But hypervisor itself may choose to use the different values which don't have to be the power of 2.

File: vmm/src/memory_manager.rs [hunk 1]
```diff
@@ -249,7 +249,8 @@ impl MemoryManager {
             if hotplug_method == HotplugMethod::VirtioMem {
                 let start_addr = GuestAddress(
                     (start_of_device_area.0 + vm_virtio::VIRTIO_MEM_DEFAULT_BLOCK_SIZE - 1)
-                        & (!(vm_virtio::VIRTIO_MEM_DEFAULT_BLOCK_SIZE - 1)),
+                        / vm_virtio::VIRTIO_MEM_DEFAULT_BLOCK_SIZE
```
Comment (teawater) [inline vmm/src/memory_manager.rs:252]: Could you change this part to use bit operation?

Comment (uMartinXu) [reply vmm/src/memory_manager.rs:252]: Your original code is right.  After I digged into the virtio-mem kernel code, I found that the block size actually doesn't have to be the power of 2 and I did not find any limitation on the block size. :)

Comment (sboeuf) [reply vmm/src/memory_manager.rs:252]: @uMartinXu so what does this mean? Are you gonna close this PR?

Comment (uMartinXu) [reply vmm/src/memory_manager.rs:252]: This PR is still needed. 
My point is this PR just restore @teawater's original implementation.

Comment (teawater) [reply vmm/src/memory_manager.rs:252]: > This PR is still needed.
> My point is this PR just restore @teawater's original implementation.

My point is this code is OK. 
But It would be better if it could use bit operation.

Comment (uMartinXu) [reply vmm/src/memory_manager.rs:252]: @teawater agree with you.
but as I mentioned above, it is easy and readable to use bit operation when block size is the power of 2. 
But I can't find a readable bit operation to handle this situation when block size can be any number.

**Comment by rbradford** [comment]
I don't see the benefit of this change. If you're going to patch the code to change the block size to non power of 2 you can also handle the fallout this creates. As we don't offer the option to change the block size I don't see the benefit of this change.

**Comment by uMartinXu** [comment]
@rbradford I am sorry that my comments may a little bit confusing. Let me clarified it straight forward. :)

The code below (this PR intends to fix) is used to calculate strat_addr which should be aligned with the block size. But the code itself is wrong. 

let start_addr = GuestAddress(start_of_device_area.0 + vm_virtio::VIRTIO_MEM_DEFAULT_BLOCK_SIZE - 1)
                        & (!(vm_virtio::VIRTIO_MEM_DEFAULT_BLOCK_SIZE - 1)),

The result is right just by accident when VIRTIO_MEM_DEFAULT_BLOCK_SIZE is 2M
Although the user can't change the block size, it is quite possible that the programmer may change it to the other value in the future. 
The bottom line is we shouldn't keep this wrong code.  :)
